tests/net_dynamic_hb.rs

```diff
@@ -0,0 +1,201 @@
+/// * `total`: Total number of nodes. Must be >= `3 * num_faulty + 1`.
+/// * `total_txs`: The total number of transactions each node will propose. All nodes will propose
+///                the same transactions, albeit in random order.
+/// * `proposals_per_epoch`: The number of transaction to propose, per epoch.
+fn do_drop_and_readd(
+    num_faulty: usize,
+    total: usize,
+    total_txs: usize,
+    proposals_per_epoch: usize,
+) {
+    let mut rng = rand::thread_rng();
+
+    // First, we create a new test network with Honey Badger instances.
+    let mut net = NetBuilder::new(0..total)
+        .num_faulty(num_faulty)
+        .message_limit(200_000)  // Limited to 200k messages for now.
+        .using(move |node| {
+            println!("Constructing new dynamic honey badger node #{}", node.id);
+            DynamicHoneyBadger::builder().build(node.netinfo)
+        }).build()
+        .expect("could not construct test network");
+
+    // We will use the first correct node as the node we will remove from and re-add to the network.
+    // Note: This should be randomized using proptest.
+    let pivot_node_id: usize = net
+        .correct_nodes()
+        .nth(0)
+        .expect("expected at least one correct node")
+        .id()
+        .clone();
+    println!("Will remove and readd node #{}", pivot_node_id);
+
+    // We generate a list of transaction we want to propose, for each node. All nodes will propose
+    // a number between 0..total_txs, chosen randomly.
+    let mut queues: collections::BTreeMap<_, Vec<usize>> = net
+        .nodes()
+        .map(|node| (node.id().clone(), (0..total_txs).collect()))
+        .collect();
+
+    // For each node, select transactions randomly from the queue and propose them.
+    for (id, queue) in queues.iter_mut() {
+        let proposal =
+            rand::seq::sample_slice(&mut rng, queue.as_slice(), proposals_per_epoch).to_vec();
+        println!("Node {:?} will propose: {:?}", id, proposal);
+
+        // The step will have its messages added to the queue automatically, we ignore the output.
+        let _ = net
+            .send_input(*id, Input::User(proposal))
+            .expect("could not send initial transaction");
+    }
+
+    // Afterwards, remove a specific node from the dynamic honey badger network.
+    net.broadcast_input(&Input::Change(Change::Remove(pivot_node_id)))
+        .expect("broadcasting failed");
+
+    // We are tracking (correct) nodes' state through the process by ticking them off individually.
+    let mut awaiting_removal: collections::BTreeSet<_> =
+        net.correct_nodes().map(|n| n.id().clone()).collect();
+    let mut awaiting_addition: collections::BTreeSet<_> =
+        net.correct_nodes().map(|n| n.id().clone()).collect();
+    let mut expected_outputs: collections::BTreeMap<_, collections::BTreeSet<_>> = net
+        .correct_nodes()
+        .map(|n| (n.id().clone(), (0..10).into_iter().collect()))
```

Comment:
Should this be `(0..total_txs).into_iter().collect()`? And is the `into_iter()` necessary?